overlord/devicestate/systems.go

```diff
@@ -176,3 +186,212 @@ func seededSystemFromModeenv() (*seededSystem, error) {
+			Path: info.MountFile(),
+		})
+		modelSnaps[info.MountFile()] = info
+		return nil
+	}
+
+	// snapd is implicitly needed
+	const snapdIsEssential = true
+	if err := getModelSnap("snapd", snapdIsEssential, ""); err != nil {
+		return nil, "", err
+	}
+	for _, sn := range model.EssentialSnaps() {
+		const essential = true
+		if err := getModelSnap(sn.SnapName(), essential, ""); err != nil {
+			return nil, "", err
+		}
+	}
+	for _, sn := range model.SnapsWithoutEssential() {
+		const essential = false
+		if err := getModelSnap(sn.SnapName(), essential, sn.Presence); err != nil {
+			return nil, "", err
+		}
+	}
+	if err := w.SetOptionsSnaps(optsSnaps); err != nil {
+		return nil, "", err
+	}
+
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		fromDB := func(ref *asserts.Ref) (asserts.Assertion, error) {
+			return ref.Resolve(db.Find)
+		}
+		return asserts.NewFetcher(db, fromDB, save)
+	}
+	f, err := w.Start(db, newFetcher)
+	if err != nil {
+		return nil, "", err
+	}
+	// past this point the system directory is present
+
+	localSnaps, err := w.LocalSnaps()
+	if err != nil {
+		return nil, recoverySystemDir, err
+	}
+
+	for _, sn := range localSnaps {
+		info, ok := modelSnaps[sn.Path]
+		if !ok {
+			return nil, recoverySystemDir, fmt.Errorf("internal error: no snap info for %q", sn.Path)
+		}
+		// TODO: the side info derived here can be different from what
+		// we have in snap.Info, but getting it this way can be
+		// expensive as we need to compute the hash, try to find a
+		// better way
+		_, aRefs, err := seedwriter.DeriveSideInfo(sn.Path, f, db)
+		if err != nil {
+			if !asserts.IsNotFound(err) {
+				return nil, recoverySystemDir, err
+			} else if info.SnapID != "" {
+				// snap info from state must have come
+				// from the store, so it is unexpected
+				// if no assertions for it were found
+				return nil, recoverySystemDir, fmt.Errorf("internal error: no assertions for asserted snap with ID: %v", info.SnapID)
+			}
+		}
+		if err := w.SetInfo(sn, info); err != nil {
+			return nil, recoverySystemDir, err
+		}
+		sn.ARefs = aRefs
+	}
+
+	if err := w.InfoDerived(); err != nil {
+		return nil, recoverySystemDir, err
+	}
+
+	for {
+		// get the list of snaps we need in this iteration
+		toDownload, err := w.SnapsToDownload()
+		if err != nil {
+			return nil, recoverySystemDir, err
+		}
+		// which should be empty as all snaps should be accounted for
+		// already
+		if len(toDownload) > 0 {
+			which := make([]string, 0, len(toDownload))
+			for _, sn := range toDownload {
+				which = append(which, sn.SnapName())
+			}
+			return nil, recoverySystemDir, fmt.Errorf("internal error: need to download snaps: %v", strings.Join(which, ", "))
+		}
+
+		complete, err := w.Downloaded()
+		if err != nil {
+			return nil, recoverySystemDir, err
+		}
+		if complete {
+			logger.Noticef("snap processing complete")
+			break
+		}
+	}
+
+	for _, warn := range w.Warnings() {
```

Comment:
the only warning from writer itself that I know of atm is UC16/18 only but this is ok